Refactor magic methods in WP_List_Table, WP_User_Query, WP_Text_Diff_Renderer_Table and WP_Admin_Bar classes

anton-vlasenko opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

This PR aims to refactor magic methods in WP_List_Table, WP_User_Query, WP_Text_Diff_Renderer_Table, and WP_Admin_Bar classes to make them compatible with PHP 8.2.  Also, it ensures that developers will get a warning when using dynamic properties on these classes.

Trac ticket: https://core.trac.wordpress.org/ticket/56876
Trac ticket: https://core.trac.wordpress.org/ticket/58896

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

anton-vlasenko committed:
WP_Admin_Bar::menu property can be removed as it's not used anywhere.

anton-vlasenko committed:
WP_Admin_Bar::__get should be removed.
It creates issues.
WP_Admin_Bar::$menu property should be declared instead to avoid any class behavior inconsistency.
For example, isset($wp_admin_bar->menu) will return false, but $wp_admin_bar::$menu returns a value.

anton-vlasenko committed:
Remove WP_Admin_Bar::$menu.
It's safe to remove it as it doesn't break BC.
See https://wpdirectory.net/search/01GC9XDX4EBTTDF7G8AJNYRCJW
All the plugins check if it exists before using it.
And it doesn't exist because this property was never defined in the __isset magic method.

anton-vlasenko committed:
Add a unit test to verify that the deprecated properties (menu, proto) are not used in the WP_Admin_Bar class.

anton-vlasenko committed:
1. Fix bug in the Tests_AdminBar::test_should_not_use_dynamic_properties method.
2. Add dataprovider.
3. Add comments.

anton-vlasenko committed:
1. Update PHPDoc blocks.
2. Rename parameters in the named dataset.
3. Update comments.

anton-vlasenko committed:
Fix CS.

anton-vlasenko committed:
Fix CS.

anton-vlasenko committed:
Change visibility modifier of the data provider.

anton-vlasenko committed:
Disallow to use dynamic properties as they are deprecated in PHP 8.2.
Warn developers about it.

anton-vlasenko committed:
Add unit tests covering predefined properties (the ones that are stored in Tests_Admin_WpListTable::$compat_fields).

anton-vlasenko committed:
Add unit tests to cover getters and setters.

anton-vlasenko committed:
Fix CS.

anton-vlasenko committed:
Update PHPDoc blocks.

jeffpaul commented:
@SergeyBiryukov @jrfnl pinging you here in hopes one of you can help give this a review and help it land in 6.1.1... thank you!

jeffpaul commented:
@hellofromtonya also pinging you as well, any help with review/commit on this for 6.1.1 would be appreciated... thank you!

anton-vlasenko committed:
Fix @covers.

anton-vlasenko committed:
The __set method must not return any values.

anton-vlasenko commented:
~~I want to add a fix for another class to this PR because the fixes are similar.
So please don't review for now.~~
The PR is ready for review.

anton-vlasenko committed:
Warn a developer if they try to set/get dynamic properties on the WP_User_Query class.

anton-vlasenko committed:
WP_User_Query::__isset should return a boolean value.

anton-vlasenko committed:
Commit WIP.

anton-vlasenko committed:
Fix the tests for dynamic properties.

anton-vlasenko committed:
Fix CS.

anton-vlasenko committed:
Don't return values in the __set method.
__isset should always return a boolean value.

anton-vlasenko committed:
Declare these properties as public, and remove the magic methods as they not needed anymore.

anton-vlasenko committed:
Revert the changes.
Let's not remove the magic methods.

anton-vlasenko committed:
Add doing_it_wrong errors to warn developers about dynamic properties.

anton-vlasenko committed:
Make linter fail (for testing purposes).

anton-vlasenko committed:
Make linter fail (for testing purposes).

anton-vlasenko committed:
Revert the debug code.

anton-vlasenko committed:
Create a new test file for WpTextDiffRendererTable.

anton-vlasenko committed:
Commit the draft version of a class.

anton-vlasenko committed:
1. Fix names of the variables.

anton-vlasenko committed:
Require classes at the header to ensure it works on all PHPUnit versions.
The firing order of dataProviders vs. set_up_before_class() seems to depend on the PHPUnit version.

audrasjb commented on src/wp-admin/includes/class-wp-list-table.php:
```diff
@@ -184,6 +184,16 @@ public function __get( $name ) {
+			__METHOD__,
+			sprintf(
+				// translators: 1: The name of the non-existent class property.
+				__( 'The "%1$s" property isn\'t defined. Dynamic properties are deprecated in PHP 8.2 and above.' ),
```
Replace `isn\'t` with `is not`.

anton-vlasenko replied:
Thank you for the review!
Fixed in 42ba291e72e740a6cd4121e78089dccd2a578c35

audrasjb commented on src/wp-admin/includes/class-wp-list-table.php:
```diff
@@ -193,12 +203,22 @@ public function __get( $name ) {
+			__METHOD__,
+			sprintf(
+				// translators: 1: The name of the non-existent class property.
+				__( 'The "%1$s" property isn\'t defined. Dynamic properties are deprecated in PHP 8.2 and above.' ),
```
Replace `isn\'t` with `is not`.

anton-vlasenko replied:
Fixed in 42ba291e72e740a6cd4121e78089dccd2a578c35

audrasjb commented on src/wp-includes/class-wp-text-diff-renderer-table.php:
```diff
@@ -506,6 +506,16 @@ public function __get( $name ) {
+			__METHOD__,
+			sprintf(
+			// translators: 1: The name of the non-existent class property.
+				__( 'The "%1$s" property isn\'t defined. Dynamic properties are deprecated in PHP 8.2 and above.' ),
```
Replace `isn\'t` with `is not`.

anton-vlasenko replied:
Fixed in 42ba291e72e740a6cd4121e78089dccd2a578c35

audrasjb commented on src/wp-includes/class-wp-text-diff-renderer-table.php:
```diff
@@ -515,12 +525,22 @@ public function __get( $name ) {
+				__( 'The "%1$s" property isn\'t defined. Dynamic properties are deprecated in PHP 8.2 and above.' ),
+				$name
+			),
+			'6.3.0'
```
Use 6.2.0

anton-vlasenko replied:
Good catch.
Fixed in a9e364eaf792b85ba0bb611f6642e40e033b9a26

audrasjb commented on src/wp-includes/class-wp-user-query.php:
```diff
@@ -1042,6 +1042,16 @@ public function __get( $name ) {
+			__METHOD__,
+			sprintf(
+			// translators: 1: The name of the non-existent class property.
+				__( 'The "%1$s" property isn\'t defined. Dynamic properties are deprecated in PHP 8.2 and above.' ),
```
Replace `isn\'t` with `is not`.

anton-vlasenko replied:
Fixed in 42ba291e72e740a6cd4121e78089dccd2a578c35

audrasjb commented on src/wp-includes/class-wp-user-query.php:
```diff
@@ -1051,12 +1061,22 @@ public function __get( $name ) {
+			__METHOD__,
+			sprintf(
+			// translators: 1: The name of the non-existent class property.
+				__( 'The "%1$s" property isn\'t defined. Dynamic properties are deprecated in PHP 8.2 and above.' ),
```
Replace `isn\'t` with `is not`.

anton-vlasenko replied:
Fixed in 42ba291e72e740a6cd4121e78089dccd2a578c35

audrasjb reviewed (changes requested):
I added a few very small change requests.

audrasjb commented:
Hi there, thanks for the PR!
I added a few very small change requests.

anton-vlasenko commented on src/wp-includes/class-wp-admin-bar.php:
```diff
@@ -18,23 +18,6 @@ class WP_Admin_Bar {
-	 * @return string|array|void
-	 */
-	public function __get( $name ) {
-		switch ( $name ) {
```
I've researched if these properties are being used (including WP plugins), and it looks like they can be safely removed without breaking BC.

hellofromtonya replied:
@anton-vlasenko is there a search query you used to investigate plugins and themes? If no, how did you investigate usage?

anton-vlasenko replied:
That's a good question, @hellofromtonya.
Unfortunately, I'm not writing down every search query that I use (if we are talking about wpdirectory).
I committed these changes sometime last year, so I can't recall.
The best I can do is to search again and post the queries here.

anton-vlasenko replied:
> I'm not writing down every search query that I use 

I was wrong, @hellofromtonya . I added the search query to a [commit message](https://github.com/anton-vlasenko/wordpress-develop/commit/7b4a22bc1e42efce267ecf76d1b096de5a89b2c4). 

So, I've used these search queries to find usages of `$proto` and `$menu` properties:
https://wpdirectory.net/search/01GC9XDX4EBTTDF7G8AJNYRCJW
https://wpdirectory.net/search/01GSTP3YTZP7R6PSHAZ51YPQBD

I was not able to find any occurrences of `WP_Admin_Bar::$proto`.
As for the `WP_Admin_Bar::$menu`, there are some occurrences, but `isset()` is always used to check if the property exists. So it should be safe to remove the `$menu` property.

anton-vlasenko replied:
@hellofromtonya 
I've changed my mind and brought `WP_Admin_Bar::$menu` back.
There is no reliable way to check all the plugins and themes because there are too many search results.
https://wpdirectory.net/search/01GSTW1X69TBN8FH3SY7V8KPY5 returns only a few results, all of which seem good (meaning that `WP_Admin_Bar::$menu` can be removed).
But if I use slightly broader search criteria (https://wpdirectory.net/search/01GSTW68SM7MNRZD0Z9AVBH6R0) I get about 4800 results. It's unrealistic to check all or at least the most popular plugins/themes.
So, to stay safe, I'd prefer to keep the `WP_Admin_Bar::$menu` property for now.
At the same time, `WP_Admin_Bar::$proto` can be removed. It doesn't seem to be used anywhere.

hellofromtonya replied:
Thank you for the search queries @anton-vlasenko. 

Reviewing the first query for `menu` shows that at least 1 plugin (s2 Member Framework) is using this property.

As [discussed in the livestream](https://www.youtube.com/live/vDZWepDQQVE?feature=share&t=9362), the plan is:

* Remove the `__get()`.
* Since `menu` is being used, add it as a public property on the class with a default value of `array()`.
```php
	/**
	 * Deprecated menu property.
	 *
	 * @since 3.1.0
	 * @deprecated 3.3.0 Modify admin bar nodes with WP_Admin_Bar::get_node(), WP_Admin_Bar::add_node(), and WP_Admin_Bar::remove_node().
	 * @var array
	 */
	public $menu = array();
```
* What about `proto` (for protocol) property? 

From the livestream, there was consensus to remove it if _unused_, but with an action to research what the property is and why it was added. 

History on the `proto` property:
* [r16038](https://core.trac.wordpress.org/changeset/16038) added this property.
* [r19501](https://core.trac.wordpress.org/changeset/19501) moved it from the `initialize()` method to `__get()`.
* When 3.1 released, the protocol handling switched to use `get_admin_url()`, `admin_url()`, etc functions instead of the `proto` property. But the property remained in the code, unused.

Would removing it be a BC break? If not used, then no. But if some plugin or theme in the wild uses it, then yes it is a BC break.

Let's remove it with this reasoning:
* It's unused in Core and none found in wporg search
* It was introduced in the 3.1 cycle but then replaced with URL get functions during that same cycle.
* It's reasonable then to say that it should have been removed in 3.1, but was missed. It appears to be dead code.

anton-vlasenko replied:
I've just submitted a [PR](https://github.com/WordPress/wordpress-develop/pull/4133) that implements the changes proposed in the @hellofromtonya 's comment above. Thanks!
https://github.com/WordPress/wordpress-develop/pull/4133

anton-vlasenko committed:
1. Replace isn't with "is not".
2. Always return value in __get() as this is what is expected from that magic methods.

anton-vlasenko committed:
Changed 6.3.0 to 6.2.0 (see https://github.com/WordPress/wordpress-develop/pull/3535/files#r1105076672).

anton-vlasenko committed:
Rollback the changes and bring WP_Admin_Bar::$menu back.
It's not possible (or very problematic) to find out whether other plugins use this property in a safe manner (i.e., checking if it exists with (isset) before using it).

anton-vlasenko committed:
Revert the changes in WP_Admin_Bar as this changes have been moved to https://github.com/WordPress/wordpress-develop/pull/4133

anton-vlasenko committed:
Revert the changes in WP_Admin_Bar as this changes have been moved to https://github.com/WordPress/wordpress-develop/pull/4133

hellofromtonya commented:
Closing this PR. Instead each of these classes will get a separate PR and commit to encapsulate their contextual history and individual considerations.
